feat(txt-registry)!: deprecate legacy txt-format

ivankatliarchuk opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

<!-- Please provide a summary of the change here. -->

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->

- follow-up for https://github.com/kubernetes-sigs/external-dns/pull/4946
- relates https://github.com/kubernetes-sigs/external-dns/pull/4624
- relates https://github.com/kubernetes-sigs/external-dns/issues/3757

TODO:
- ✅ fix tests
- ✅ test on real cluster and account
- ✅ review docs
- ✅ added TXT legacy record cleanup script

**Checklist**

- [X] Unit tests updated
- [X] End user documentation updated


Executed on real cluster with arguments

```sh
go run main.go \
    --provider=aws \
    --registry=txt \
    --source=fake \
    --aws-zone-type=private \
    --zone-id-filter=/hostedzone/XXXXXX \
    --log-level=debug \
    --policy=sync \
    --interval=60s \
    --fqdn-template=a1.ex.com
```

Without change, records current and old format created

```
aws route53 list-resource-record-sets --hosted-zone-id ${ZONE_UNDER_TEST} --query "ResourceRecordSets[?Type=='TXT'].{Name:Name, Value:ResourceRecords[0].Value}" --output table
-----------------------------------------------------------------------------
|                          ListResourceRecordSets                           |
+--------------------+------------------------------------------------------+
|        Name        |                        Value                         |
+--------------------+------------------------------------------------------+
|  a-fhbr.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-gwbl.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-hgmc.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-hmqw.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-hoxg.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-ipfd.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-melr.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-mkde.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-uknh.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-xiqj.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  fhbr.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  gwbl.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  hgmc.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  hmqw.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  hoxg.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  ipfd.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  melr.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  mkde.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  uknh.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  xiqj.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
+--------------------+------------------------------------------------------+
```

With the change. New format records created, letagcy TXT records left untouched

```
❯❯ aws route53 list-resource-record-sets --hosted-zone-id ${ZONE_UNDER_TEST} --query "ResourceRecordSets[?Type=='TXT'].{Name:Name, Value:ResourceRecords[0].Value}" --output table
-----------------------------------------------------------------------------
|                          ListResourceRecordSets                           |
+--------------------+------------------------------------------------------+
|        Name        |                        Value                         |
+--------------------+------------------------------------------------------+
|  a-dkmw.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-ihkf.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-iztb.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-jlmh.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-kmam.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-lqpv.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-rizx.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-sjfa.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-uvlq.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  a-vgpc.a1.ex.com. |  "heritage=external-dns,external-dns/owner=default"  |
|  bwwo.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  ecvx.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  ilan.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  iyda.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  nhhy.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  pazy.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  sdeo.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  tosb.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  wrej.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
|  yymt.a1.ex.com.   |  "heritage=external-dns,external-dns/owner=default"  |
+--------------------+------------------------------------------------------+
``

ivankatliarchuk commented:
/label tide/merge-method-squash

ivankatliarchuk commented:
@mloiseleur , @mcharriere  wdyt?

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -5,6 +5,24 @@ It stores DNS record metadata in TXT records, using the same provider.
+
+Occasionally, it may be necessary to remove outdated TXT records from your registry.
+
+The script could be found in [scripts/cleanup-legacy-txt-records.py](../../scripts/cleanup-legacy-txt-records.py) with instructions how to run it.
```
```suggestion
An exaloke script can be found in [scripts/cleanup-legacy-txt-records.py](../../scripts/cleanup-legacy-txt-records.py) with instructions on how to run it.
```

mloiseleur commented on pkg/apis/externaldns/types.go:
```diff
@@ -615,7 +615,7 @@ func App(cfg *Config) *kingpin.Application {
 	app.Flag("txt-encrypt-enabled", "When using the TXT registry, set if TXT records should be encrypted before stored (default: disabled)").BoolVar(&cfg.TXTEncryptEnabled)
 	app.Flag("txt-encrypt-aes-key", "When using the TXT registry, set TXT record decryption and encryption 32 byte aes key (required when --txt-encrypt=true)").Default(defaultConfig.TXTEncryptAESKey).StringVar(&cfg.TXTEncryptAESKey)
-	app.Flag("txt-new-format-only", "When using the TXT registry, only use new format records which include record type information (e.g., prefix: 'a-'). Reduces number of TXT records (default: disabled)").BoolVar(&cfg.TXTNewFormatOnly)
+	app.Flag("txt-new-format-only", "A key is reserved for TXT migraion. Enabling or disabling it will not take any effect. (deprecated)").BoolVar(&cfg.TXTNewFormatOnly)
```
```suggestion
	app.Flag("txt-new-format-only", "Was used during TXT format migration. Enabling or disabling it will not take any effect. (deprecated)").BoolVar(&cfg.TXTNewFormatOnly)
```

mloiseleur commented on scripts/cleanup-legacy-txt-records.py:
```diff
@@ -0,0 +1,213 @@
+#    - The script is executed with command-line arguments specifying the hosted zone ID, record pattern, total items to delete, batch size, and whether to perform a dry run or actual deletion.
+#    - Check 'To Run script' section for more details
+
+# To Run script
+# 1. Python, pip and pipenv installed https://pipenv.pypa.io/en/latest/
+# 2. AWS Access https://docs.aws.amazon.com/signin/latest/userguide/command-line-sign-in.html
+# 3. pipenv shell
+# 4. pip install boto3
+# 5. python scripts/cleanup-legacy-txt-records.py --help
+# WARNING: run this script at your own RISK. This will delete all the TXT rerods that do contain certain string.
```
```suggestion
# WARNING: run this script at your own RISK. This will delete all the TXT rerods that do contain certain string.
# To Run script
# 1. Python, pip and pipenv installed https://pipenv.pypa.io/en/latest/
# 2. AWS Access https://docs.aws.amazon.com/signin/latest/userguide/command-line-sign-in.html
# 3. pipenv shell
# 4. pip install boto3
# 5. python scripts/cleanup-legacy-txt-records.py --help
```

ivankatliarchuk commented:
/retest

mloiseleur commented:
/lgtm

Evesy commented:
It feels wrong to be continuing forward moving to the newer TXT format and removing the legacy format given the [issues](https://github.com/kubernetes-sigs/external-dns/issues/3757) with the new format.

Right now with the dual TXT setup, assuming a user uses a `--txt-prefix` subdomain (i.e. `--txt-prefix=myprefix.`), they can successfully manage apex records.

For `example.com` the old format creates a TXT record `myprefix.example.com` (Success), and the new format will constantly try to create e.g. `myprefix.cname-example.com` (Failure), which doesn't work as that domain is unlikely to be owned.

Moving to the new TXT format only removes external DNS's ability to be able to manage APEX domains at all in this scenario, instead you'll have the following:

- First time external-dns creates the normal DNS successfully, and fails to create the corresponding TXT record
- Second time round, external-dns sees the record already exists with no matching ownership TXT record, so it assumes it does not manage the record

ivankatliarchuk commented:
This could sound strange or something, but if APEX domain management is required or if someone would like to stay on legacy format, one of the options is to stay on an external-dns version that supports required functionality. Having round-tricps and backoff not make much sense, the logic is going to be quite hard to understand.

I have no issue to close this pull request. The new format does not work for me, same as legacy. So the idea is to kill this code and ask where the owner of a pull-request https://github.com/kubernetes-sigs/external-dns/pull/4624 would like to proceed with his changes.

Another option is to support CRD registry

ivankatliarchuk commented:
Apex support is buggy.... https://github.com/kubernetes-sigs/external-dns/issues/449. I manage APEX in terraform,own code at the moment.

ivankatliarchuk commented:
Apex support is buggy.... https://github.com/kubernetes-sigs/external-dns/issues/449. I manage APEX in terraform, or have some propriatery code at the moment.

ivankatliarchuk commented:
> For example.com the old format creates a TXT record myprefix.example.com (Success), and the new format will constantly try to create e.g. myprefix.cname-example.com (Failure), which doesn't work as that domain is unlikely to be owned.

This looks like a bug to me for when custom prefix is specified. At least for APEx it should behave differently

ivankatliarchuk commented:
> For example.com the old format creates a TXT record myprefix.example.com (Success), and the new format will constantly try to create e.g. myprefix.cname-example.com (Failure), which doesn't work as that domain is unlikely to be owned.

This looks like a bug to me for when custom prefix is specified. At least for APEx it should behave differently. Could you share all you arguments and example setup. Maybe there is a way to write a fix

Evesy commented:
I agree, the TXT registry has been a bit of a mess for a while now, ever since the new format was introduced to help support managing the same record with multiple record types. It solved that one problem but introduced a number of other compatibility problems for users along with it. In my opinion it should have been removed back then until a proper working solution was found, but instead it's been in a halfway house for years now.

I do think the newer registry format works for all use cases, including apex domains, where templating is being used, i.e. for my example above instead of `--txt-prefix=myprefix.` you would use `--txt-prefix=myprefix-%{record_type}.`. So perhaps what's really needed is an inbuilt support mechanism in external-dns for transferring from one prefix to another. This way anyone using prefixes/suffixes from before the new registry format was added would have a migration path between old and new config?

ivankatliarchuk commented:
So how it currently works for Apex domains? I have not found a flag to disable upsert for new format. Does external-dns errors out, crashes, or swallow the error?

Evesy commented:
Our setup has the below flags:
```
        - --source=ingress
        - --source=service
        - --source=crd
        - --provider=cloudflare
        - --policy=sync
        - --registry=txt
        - --txt-owner-id=dpp
        - --annotation-filter=external-dns.alpha.kubernetes.io/ignore notin (true)
        - --crd-source-apiversion=externaldns.k8s.io/v1alpha1
        - --txt-prefix=prefix.
 ```

What we see in the case of apex domains is external-dns constantly updating the records no every reconciliation loop:
```
time="2025-03-18T07:32:16Z" level=debug msg="Skipping record myprefix.cname-apex.co.uk because no hosted zone matching record DNS Name was detected"
time="2025-03-18T07:32:31Z" level=info msg="Changing record." action=UPDATE record=apex.co.uk ttl=1 type=CNAME zone=2db0e2d6e06b42e72e5479c062f3a3de
time="2025-03-18T07:32:31Z" level=info msg="Changing record." action=UPDATE record=myprefix.apex.co.uk ttl=1 type=TXT zone=2db0e2d6e06b42e72e5479c062f3a3de
```

My assumption, and I'm sure I confirmed a while ago, is that external-dns is reconciling everytime because on each check it sees the `${prefix}-${record_type}.` TXT ownership record is missing. Fortunately, the DNS record for the apex is still created and managed by external-dns, because the legacy TXT format is successfully created and shows ownership.

However, if external-dns begins only caring about `${prefix}-${record_type}.` prefix type, this does not work for apex domains with the above configuration; the record can't be created and so external-dns will be unable to manage the domain

ivankatliarchuk commented:
Got it. Will try to find a solution for apex records, before merging this to master

/hold

ivankatliarchuk commented:
I think the question is, how to reliably `determine if zone apex or subdomain`.......

Evesy commented:
> I think the question is, how to reliably `determine if zone apex or subdomain`.......

The way we've handled a similar problem elsewhere (determining the authoritative zone for a domain) is to traverse through the address until an SOA record is returned, e.g.

```
$ dig SOA app.subdomain.tld.co.uk
// No SOA

$ dig SOA subdomain.tld.co.uk
// No SOA

$ dig SOA tld.co.uk
// SOA, so authorative
```

The same logic could be applied; if the record being created resolves an SOA, it's probably an apex domain

Though I'm not sure external-dns should be doing this as it'd be a lot of extra lookups and might not be fool proof in all scenarios, especially where split DNS is involved. I do think the better approach is to a) clearly document the downfalls of using `--txt-prefix` when _not_ using the interpolated `%{record_type}` (arguably you could enforce that `%{record_type}` is set within `txt-prefix`, and b) provide a migration for users to be able to change `txt-prefix` and have external-dns migrate the existing TXT entries automatically

ivankatliarchuk commented:
DNS lookups, not sure, most likely not an option.  We all have different networking requirements, network segmentation, what if there is no access to DNS lookups and etc.

Maybe we need a new flag `--apex-domains` and-or `--apex-domains-regex`?

ivankatliarchuk commented:
DNS lookups, not sure, most likely not an option.  We all have different networking requirements, network segmentation, what if there is no access to DNS lookups and etc.

Maybe we need a new flag `--apex-domains` and-or `--apex-domains-regex`?

And add a flag `--allow-dns-lookup` ? As we already do in service ? https://github.com/kubernetes-sigs/external-dns/blob/ecd57c86f526aebf1ecf54d0291de61ec3d48d6d/source/service.go#L546

ivankatliarchuk commented:
DNS lookups, not sure, on the edge.  We all have different networking requirements, network segmentation, what if there is no access to DNS lookups and etc.

Maybe we need a new flag `--apex-domains` and-or `--apex-domains-regex`?

And add a flag `--allow-dns-lookup` ? As we already do for service ? https://github.com/kubernetes-sigs/external-dns/blob/ecd57c86f526aebf1ecf54d0291de61ec3d48d6d/source/service.go#L546

Evesy commented:
I suppose even if you could reliably determine a domain is an apex, what would external-dns do to handle the domain differently? 

You could suffix it with a period `.` (unless already the case) to ensure the TXT record is created as a subdomain of the apex. This is probably unlikely to have any compatibility issues that I can think of but as you say, people have very different setups.

Not everyone uses it, but `--domain-filter` could be used as a known list of apex zones if present (Admittedly haven't given  much thought of the implications of this, so would need to think)

I'm still in favour of consistently handling records, rather than having edge cases for apex domains where external-dns exhibits different behaviour; and to achieve that I personally feel either mandating or clearly warning users about the impacts of using `--txt-prefix` when it comes to apex domains (i.e. If you want to manage apex domains, you should ensure the txt-prefix ultimately becomes a subdomain (the same problem surely applies with txt-suffix too)) is the best approach, with, ideally, a migration path for users

ivankatliarchuk commented:
Got it. I'll submit a patch with documentation update. Yeah, handling apex corretly is not a trivial thing. The dns lookup could happen for example for domain `test.one.example.co.uk` from right to left kinda `lookup uk`, `lookup co.uk`, `lookup example.co.uk BINGO` but yeah, this sort of over engineering things.

Will update docs and flag description

ivankatliarchuk commented:
So, the behaviour you described, could be a source or provider specific, as I was not able to reproduce it for AWS provider. There is an open issue from 2018.... https://github.com/kubernetes-sigs/external-dns/issues/449

so this are my fixtures, where apex zone is `ex.com`

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: extdns
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-v1
  namespace: extdns
  annotations:
    dns.company.com/label: controllertest-v1
    dns.issue/type: issue-5172
    external-dns.alpha.kubernetes.io/hostname: nginx-v1.ex.com
spec:
  type: ClusterIP
  clusterIP: None
  ports:
    - port: 80
      name: http
      targetPort: 80
  selector:
    app: nginx
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-v2
  namespace: extdns
  annotations:
    dns.company.com/label: controllertest-v2
    dns.issue/type: issue-5172
    external-dns.alpha.kubernetes.io/ttl: "1m"
    external-dns.alpha.kubernetes.io/hostname: nginx-v2.ex.com
spec:
  type: LoadBalancer
  allocateLoadBalancerNodePorts: true
  ports:
    - port: 80
      name: http
      targetPort: 80
  selector:
    app: nginx
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-v3
  namespace: extdns
  annotations:
    dns.company.com/label: controllertest-v2
    dns.issue/type: issue-5172
    external-dns.alpha.kubernetes.io/ttl: "180"
    external-dns.alpha.kubernetes.io/hostname: nginx-v3.ex.com
spec:
  type: LoadBalancer
  allocateLoadBalancerNodePorts: true
  ports:
    - port: 80
      name: http
      targetPort: 80
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: extdns
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - image: nginx
          name: nginx
          ports:
            - containerPort: 80
              name: http
```

and arguments for latest external dns

```sh
go run main.go \
    --provider=aws \
    --registry=txt \
    --source=service \
    --aws-zone-type=private \
    --zone-id-filter=/hostedzone/${ZONE} \
    --log-level=debug \
    --policy=sync \
    --interval=10s \
    --txt-owner-id="current" \
    --annotation-filter="dns.issue/type=issue-5172"
```

results

```
  ❯❯ aws route53 list-resource-record-sets \
        --hosted-zone-id ${ZONE} \
        --query "ResourceRecordSets[?Type=='TXT'].{Name:Name, Value:ResourceRecords[0].Value}" --output table
----------------------------------------------------------------------------------------------------------------------------
|                                                  ListResourceRecordSets                                                  |
+--------------------+-----------------------------------------------------------------------------------------------------+
|        Name        |                                                Value                                                |
+--------------------+-----------------------------------------------------------------------------------------------------+
|  a-nginx-v2.ex.com.|  "heritage=external-dns,external-dns/owner=current,external-dns/resource=service/extdns/nginx-v2"   |
|  a-nginx-v3.ex.com.|  "heritage=external-dns,external-dns/owner=current,external-dns/resource=service/extdns/nginx-v3"   |
+--------------------+-----------------------------------------------------------------------------------------------------+
```

and for arguments
```sh
go run main.go \
    --provider=aws \
    --registry=txt \
    --source=service \
    --aws-zone-type=private \
    --zone-id-filter=/hostedzone/${ZONE} \
    --log-level=debug \
    --policy=sync \
    --interval=10s \
    --txt-owner-id="current" \
    --annotation-filter="dns.issue/type=issue-5172" \
    --txt-prefix="%{record_type}-prefix-"
```

and results
```sh
❯ aws route53 list-resource-record-sets --hosted-zone-id ${ZONE} --query "ResourceRecordSets[?Type=='TXT'].{Name:Name, Value:ResourceRecords[0].Value}" --output table
-----------------------------------------------------------------------------------------------------------------------------------
|                                                     ListResourceRecordSets                                                      |
+---------------------------+-----------------------------------------------------------------------------------------------------+
|           Name            |                                                Value                                                |
+---------------------------+-----------------------------------------------------------------------------------------------------+
|  a-prefix-nginx-v2.ex.com.|  "heritage=external-dns,external-dns/owner=current,external-dns/resource=service/extdns/nginx-v2"   |
|  a-prefix-nginx-v3.ex.com.|  "heritage=external-dns,external-dns/owner=current,external-dns/resource=service/extdns/nginx-v3"   |
+---------------------------+-----------------------------------------------------------------------------------------------------+
```

ivankatliarchuk commented:
I've updated documentation. In regards `external-dns` to provide a migration path, there is a note that was there. I have tried to implement a migration, but there are no easy way to implement that. Current behaviour, the `external-dns` will create a new TXT records and leave legacy records untouched for cleanup. The script to run `targeted` deletion of TXT records is already added, not an ideal, but should help a bit.

> Note: `external-dns` will not automatically remove legacy format records when switching to new-format-only mode. You'll need to clean up the old records manually if desired.

If you ok with following changes, submit LGTM or share what is missing so I could `un-hold` the pull request

Evesy commented:
> So, the behaviour you described, could be a source or provider specific, as I was not able to reproduce it for AWS provider.

The problem exists when a user is specifying a `txt-prefix` that does not include the `%{record_type}` substitution, and the prefix ends in a subdomain, i.e. `myprefix.`

So when trying to create a record for i.e. tld.co.uk, the old TXT record format creates correctly (TXT myprefix.tld.co.uk), however the new TXT record format (TXT myprefix.a-tld.co.uk) does not, since it's trying to create a record for an entirely different domain.

ivankatliarchuk commented:
I just shared my configuration, and there is no problem I discover. Running on latest external-dns. The only different is AWS Route53, not cloudflare.

ivankatliarchuk commented:
Have a look https://github.com/kubernetes-sigs/external-dns/pull/5172#issuecomment-2735479292. For apex domain, without even having a prefix, the new TXT record is `a-nginx-v2.ex.com`. There is no subdomain created.

ivankatliarchuk commented:
I updated the docs. Personally not sure if docs update even need to be required, as not able to reproduce the behaviour in my setup, but no harm to have it there.

ivankatliarchuk commented:
@mloiseleur I could be missing something obvious for apex domains. wdyt?

Evesy commented:
@ivankatliarchuk If you try with `--txt-prefix=foo.` you should be able to see the issue I would imagine

ivankatliarchuk commented:
Understood. But this is a configuration that not default, so one could just break things. For example, if prefix contains IDNA character, example `點` this most likely will cripple/break records as well. And most of the providers do not even support cases. 

What I'm unsure. The expectation is, that by default the TXT records are going to be update, at least they should. There are dozens of flags that could break any TXT record format, but this on users.

Would you mind share your desired message in documentation, or even create a pull request with examples and mitigations. This could help for sure to progress with this PR as well.

Evesy commented:
The README already contains this:
> If using a txt registry and attempting to use a CNAME the --txt-prefix must be set to avoid conflicts. Changing --txt-prefix will result in lost ownership over previously created records.

I would suggest adding something like the below? 
> If you plan to manage apex domains with external-dns whilst using a txt registry, you should ensure when using `--txt-prefix` that you specify the record type substitution and  that it ends in a period `.`, to ensure the record is created under the same domain as the apex record being managed, i.e. `--txt-prefix=someprefix-%{record_type}.`

ivankatliarchuk commented:
Nice one. Added to `docs/registry/txt.md`

ivankatliarchuk commented:
/unhold

cc: @mloiseleur

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,49 @@
+|             Name              |   TYPE   |
+|:-----------------------------:|:--------:|
+|   `a-abc-nginx-v2.ex.com.`    |  `TXT`   |
+| `cname-abc-nginx-v3.ex.com.`  | `CNAME`  |
```
```suggestion
| `nginx-v2.ex.com.`  | `CNAME`  |
```
The CNAME should be without the prefix, no ? Or did I miss something ?

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,49 @@
+|             Name              |  TYPE   |
+|:-----------------------------:|:-------:|
+|   `nginx-v2-abc.a.ex.com.`    |  `TXT`  |
+| `nginx-v3-abc.cname.ex.com..` | `CNAME` |
```
```suggestion
| `nginx-v3.ex.com..` | `CNAME` |
```

mloiseleur commented on scripts/cleanup-legacy-txt-records.py:
```diff
@@ -0,0 +1,213 @@
+#    - The script is executed with command-line arguments specifying the hosted zone ID, record pattern, total items to delete, batch size, and whether to perform a dry run or actual deletion.
+#    - Check 'To Run script' section for more details
+
+# WARNING: run this script at your own RISK. This will delete all the TXT rerods that do contain certain string.
```
```suggestion
# WARNING: run this script at your own RISK. This will delete all the TXT records that do contain certain string.
```

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,49 @@
 The TXT registry is the default registry.
 It stores DNS record metadata in TXT records, using the same provider.
 
+If you plan to manage apex domains with external-dns whilst using a txt registry, you should ensure when using --txt-prefix that you specify the record type substitution and that it ends in a period ., to ensure the record is created under the same domain as the apex record being managed, i.e. --txt-prefix=someprefix-%{record_type}.
```
```suggestion
If you plan to manage apex domains with external-dns whilst using a txt registry, you should ensure when using --txt-prefix that you specify the record type substitution and that it ends in a period (**.**). The record should be created under the same domain as the apex record being managed, i.e. --txt-prefix=someprefix-%{record_type}.
```

mloiseleur commented:
@ivankatliarchuk This PR lgtm. I left a last suggestion in order to reduce sentence length (and so increase readability).

@mrozentsvayg Do you think you can review this PR ?

ivankatliarchuk commented:
Thanks. So the changes are
- added `aws-` prefix to a script, as in docs it was saying it, but nothing in the script
- removed old method `toTXTName` and renamted `toNewTXTName` to a `toTXTName`

Removing legacy TXT records is not required, and this is optional operation that is already documented

My personal opinion, TXT record type is a registry is a wrong idea on it's own. Not flexible at all.

ivankatliarchuk commented:
Could you share exact wording pls what to add to README. Also, would you be open to creating a TXT registry best practices page or section?

mrozentsvayg commented:
> Could you share exact wording pls what to add to README.

I would suggest something like:

**Note**: `--txt-prefix` and `--txt-suffix` contribute to the 63-byte maximum record length.
To avoid errors, use them only if absolutely required and keep them as short as possible.

> Also, would you be open to creating a TXT registry best practices page or section?

Sure, there is no specific/pressing timeline for that, is there?

> My personal opinion, TXT record type is a registry is a wrong idea on it's own. Not flexible at all.

Right, but we need an alternative. If you're not using AWS, the TXT registry is the only way to manage the same zone with multi-cluster and sync policies.
For cloudflare specifically, we could use the record tags. Not an ideal solution (tags are not KV, only alphanumeric, hyphens and underscores are allowed), but would work to set the record owner. For other providers no such luck, in general.

ivankatliarchuk commented:
Yeah this python script is just a helper. Up to you. Contributors could do lgtm

ivankatliarchuk commented:
This pr could w8 if you have plans to run e-2-e tests

rouke-broersma commented:
This seems like a very strange decision to me. We use external dns to manage dns records with automation, so we don't have manual maintenance of dns. External dns stores ownership information in txt records automatically. External dns migrates to a new storage format. Now users are left manually cleaning up old ownership information. This creates an enormous amount of manual cleanup for thousands of users. Externaldns can already clean up the old style txt records when the a records are deleted, why would externaldns not clean them up when the new flag introduced in https://github.com/kubernetes-sigs/external-dns/pull/4946 is enabled? Externaldns clearly knows the format of the old style txt records still.

mloiseleur commented:
> Now users are left manually cleaning up old ownership information. This creates an enormous amount of manual cleanup for thousands of users. External-dns can already clean up the old style txt records when the a records are deleted, why would external-dns not clean them up when the new flag introduced in #4946 is enabled?

This project is 100% community driven. I agree with you that would be far better if we collectively manage to reach this stage. I'm unsure if we can do it, if the implementation is really that easy. Feel free to open a PR.

rouke-broersma commented:
> > Now users are left manually cleaning up old ownership information. This creates an enormous amount of manual cleanup for thousands of users. External-dns can already clean up the old style txt records when the a records are deleted, why would external-dns not clean them up when the new flag introduced in #4946 is enabled?
> 
> This project is 100% community driven. I agree with you that would be far better if we collectively manage to reach this stage. I'm unsure if we can do it, if the implementation is really that easy. Feel free to open a PR.

I'm not saying it's that easy, I have no experience in this area of external-dns. I'm saying that there must already be a mechanism to identify and removal the old style txt records, by virtue of external-dns already doing this, so I extrapolate that it must be possible. 

I am questioning why the previous pull request was accepted without an automatic migration scenario, and I am cautioning to not accept this pull request without first adding an automatic migration, because it places a great manual maintenance burden on end users. That would be antithesis to the whole purpose of this project.

ivankatliarchuk commented:
I do not think you understand the complexity of what you are saying. There is no way to reliably and safely automatically cleanup  legacy TXT records. Just try to explain how would you do that for all providers and different combinations with prefix, with suffix, and etc. One week is enough for you to came up with the idea and examples in other languages like python, we could add them to scripts?

One of the issues with legacy and current TXT formats is that, they created with the intention to be removed manually. Just go over issues and pull requests. There is a design flow in both of the formats. Example proposal for new formats https://github.com/kubernetes-sigs/external-dns/pull/4624#issuecomment-2626978003 to have a version. Without that, the only way to cleanup TXT records now, is to remove ALL records and recreate all of the records again.

If I have a solution, it will be already out, but I do not see one, and having dangling TXT records is not a risk but inconvenience

mcharriere commented:
> Later on, the old format will be dropped and only the new format will be kept (<record_type>-<endpoint_name>).
> 
> Cleanup will be done by controller itself.

The author of the new format saw a possibility of cleaning up the old registry records here: https://github.com/kubernetes-sigs/external-dns/commit/25c7cb28ea9a0fa96a581e8770f753a35aa25f77

Although some of the code changed since then, [this line](https://github.com/kubernetes-sigs/external-dns/blame/master/registry/txt.go#L292) still suggest the clean up can be done there.
I'm playing a bit with it but there are some edge cases that need further research.

ivankatliarchuk commented:
Ok, so this is a simple case. When you have different owner ids. So here is an issue. Legacy format has one format ID, then for non legacy to have another owner ID, all records needs to be recreated...... Of course is easy to cleanup in that case but with the downtime.

ivankatliarchuk commented:
Ok, so in this case. 

When you have different owner ids. So here is an issue. Legacy format has one owner ID, then for non legacy to have another owner ID, all records needs to be recreated...... Of course is easy to cleanup in that case but with the downtime.

mcharriere commented:
> Legacy format has one owner ID, then for non legacy to have another owner ID, all records needs to be recreated

which case is that one? changing owner id is not even supported by external-dns

ivankatliarchuk commented:
I'm sorry, I'm not entirely clear on what's being asked. The issue is that I haven't discovered a reliable way to automate legacy TXT record deletions without downtimes. That's why it's not in this pull request. If someone knows how to automate this, nothing wrong to open a pull request with proposed functionality.

mrozentsvayg commented:
> The author of the new format saw the possibility of cleaning up the old registry records here: [25c7cb2](https://github.com/kubernetes-sigs/external-dns/commit/25c7cb28ea9a0fa96a581e8770f753a35aa25f77)
> 
> Although some of the code changed since then, [this line](https://github.com/kubernetes-sigs/external-dns/blame/master/registry/txt.go#L292) still suggest the clean up can be done there. I'm playing a bit with it but there are some edge cases that need further research.

this is not entirely correct, the reservation here is that the `Registry.ApplyChanges()` code only [gets called on `plan.Changes.HasChanges()`](https://github.com/kubernetes-sigs/external-dns/blob/master/controller/controller.go#L257-L258)
it's a graceful approach of deleting the records(provider agnostic), but only eventually consistent

the effective solution on my opinion could be adding specific one-off mode for the external-dns or building a separate tool to aggressively scan the txt registry regardless of the changes and:
1. delete the old-style txt records with the respect to the existing prefix/suffix/owner configuration (can't do much more for all the combinations);
2. inform if TXT records are missing for the expected DNS records (now it's also only get handled in case of changes) and potentially try to fix it;
3. inform if there are extra TXT records in the zone unexpected for the current DNS records and prefix/suffix/owner configuration;

if we would like to discuss it further, it doesn't necessarily have to block this PR, which essentially is to stop adding more garbage TXT records, especially for the new users

mcharriere commented:
> this is not entirely correct, the reservation here is that the Registry.ApplyChanges() code only [gets called on plan.Changes.HasChanges()](https://github.com/kubernetes-sigs/external-dns/blob/master/controller/controller.go#L257-L258)

the txt registry has a way to force an update (see https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go#L217); making the solution possible.

This PR deletes a part that is needed to implement the cleanup of the old-format records.

ivankatliarchuk commented:
Seems like noone suggested a solution to what was discussed here. 

Rationale to remove legacy TXT formats
- We could not or it's extremely difficult to fix bugs related to formats, and list is growing
- We are stuck with both formats and no plan

Drawback
- This PR is not going to help cleanup environments. I personally see not safe/reliable way doing that, if it exists, the community had time to implement one.

Two options, merging this PR or closing. I'm fine with any of them

@Raffo or @szuecs could you make a decision. Go or No-go?

mloiseleur commented:
/approve cancel

mloiseleur commented:
/cancel approve

mloiseleur commented:
/cancel lgtm

mloiseleur commented:
/lgtm cancel

mloiseleur commented:
Sorry, I thought I was on an other PR. I resumed this PR to its previous state.

Yunir commented:
do we want to change `cname-*` to `a-*` for AWS txt record for aliases to make it similar to ipv6 AAAA-record&`aaaa-*` txt-record to just remove confusion? (confussion discussion is going here: https://github.com/kubernetes-sigs/external-dns/issues/2903#issuecomment-2900025451) Or we can leave it just as a legacy of a product, because it does not brake anything. 
If we want to change it, then I think it's easier first to do rename-PR for TXT new format of A alias record for AWS from prefix `cname-*` to `a-*`. If no, then we are good to go with this PR.
I am also eager to reduce the amount of txt records to one.

frittentheke commented:
> do we want to change `cname-*` to `a-*` for AWS txt record for aliases to make it similar to ipv6 AAAA-record&`aaaa-*` txt-record to just remove confusion? (confussion discussion is going here: [#2903 (comment)](https://github.com/kubernetes-sigs/external-dns/issues/2903#issuecomment-2900025451)) Or we can leave it just as a legacy of a product, because it does not brake anything. If we want to change it, then I think it's easier first to do rename-PR for TXT new format of A alias record for AWS from prefix `cname-*` to `a-*`. If no, then we are good to go with this PR. I am also eager to reduce the amount of txt records to one.

Yes. Remove old format first! No option, no switch. Simply drop it from external-dns.
The migration phase has been going on long enough as incoming changes and bugfixes now actively cause issues with having two formats.

If that is done, I strongly vote for fixing the prefix for A-records to `a-`. Them being an alias is simply the technicality of AWS (or other providers which support it). The record TYPE is still either A or AAAA.

Just my 2 cents.

Yunir commented:
> Yes. Remove old format first! ... If that is done, I strongly vote for fixing the prefix for A-records to a-.

I think you misread my message. 
If we want to move from `cname-` to `a-`, then it is safer to do it **prior the removal** of old-txt format. Why:
- old-txt-format needs for external-dns to continue manage the records w/o any disruption while refactoring
- new-txt-format can be changed easily, because it's additional record for now (refactor from `cname-` to `a-`) 

But if we first remove old-txt format, then we will be left without 'intermediate' working option (it will introduce the 'breaking change' with 'action needed') - external-dns will manage either only A record which has `cname-<domain>` or `a-<domain>`, not both. It will be harder to migrate. If we do it, then external-dns will lose rights to manage all records, because it would try to search for `a-<domain>` only, but old version were creating just `cname-<domain>`.

So, I am more asking if we want to remove that confusion with prefix, then propose to do it first, as it will be much easier to do the refactor and migrate. But if we don't want to remove confusion, then we might just leave `cname-` prefix for A-alias aws records as a legacy, it is also fine.

frittentheke commented:
> > Yes. Remove old format first! ... If that is done, I strongly vote for fixing the prefix for A-records to a-.
> 
> I think you misread my message. If we want to move from `cname-` to `a-`, then it is safer to do it **prior the removal** of old-txt format. Why:
> 
>     * old-txt-format (txt w/o prefix) needs for external-dns to continue manage the records w/o any disruption while refactoring
> 
>     * new-txt-format can be changed easily, because it's additional record for now (refactor from `cname-` to `a-`)
> 
> 
> But if we first remove old-txt format, then we will be left without 'intermediate' working option (it will introduce the 'breaking change' with 'action needed') - external-dns will manage either only A record which has `cname-<domain>` or `a-<domain>`, not both. It will be harder to migrate. If we do it, then external-dns will lose rights to manage all records, because it would try to search for `a-<domain>` only, but old version were creating just `cname-<domain>`.
> 
> So, I am more asking if we want to remove that confusion with prefix, then propose to do it first, as it will be much easier to do the refactor and migrate. But if we don't want to remove confusion, then we might just leave `cname-` prefix for A-alias aws records as a legacy, it is also fine.


No I did not misread it. The opportunity to clean up anything before the new format is LONG gone. It's out there, for quite a while and the last releases actively communicated the intend to remove the old format. We pushed ahead and applied `--txt-new-format-only` in our environment and removed all legacy TXT records in the process.
See e.g. https://github.com/kubernetes-sigs/external-dns/blob/cd624b6f550c75a5accab84b1354f851e232f384/docs/registry/txt.md#record-format-options which go us started early knowing this very PR here was already open and deprecation was coming. 

In the field you will now likely encounter a wild mix of both TXT record formats.

There also was plenty discussion around any potential shortcomings in https://github.com/kubernetes-sigs/external-dns/issues/3757.

To clarify what I meant:

Remove the old TXT record format for good (making any of those TXT records just meaningless garbage to external-dns).
Then and only then(!) pick up on the renaming bugfix and please add an active conversion / rewrite of the prefix. This should be faily easy then as it's not possible to have a CNAME and and A record at the same time. So up-converting this possible.

ivankatliarchuk commented:
> But if we first remove old-txt format, then we will be left without 'intermediate' working option (it will introduce the 'breaking change' with 'action needed') - external-dns will manage either only A record which has `cname-<domain>` or `a-<domain>`, not both. It will be harder to migrate. If we do it, then external-dns will lose rights to manage all records, because it would try to search for `a-<domain>` only, but old version were creating just `cname-<domain>`.

There is a intermediate option - which is current TXT format(new format), introduced years back. Is there is an open issue that describes this problem, I could have a look, just need to know how to reproduce it.

ivankatliarchuk commented:
> If that is done, I strongly vote for fixing the prefix for A-records to `a-`. Them being an alias is simply the technicality of AWS (or other providers which support it). The record TYPE is still either A or AAAA.

I'll try to have a look at this one. I remember it was mentioned in one of the issues. Just need to find and see if there is an easy fix.

szuecs commented:
I think the AWS case `cname-` for "A" record is only partially true. For example if you have an ingress that creates an ALB, you will get an AWS specific thing called ALIAS which looks like a CNAME (target is a name), but is actually a couple of A records. The only reliable way to handle a nice name to this ALB target is to create a CNAME to the ALIAS, which is supported in R53.

From what I read in the comments, we have 2 choices:
1. merge and deletion of old ownership records is done by the community. They or their providers can share scripts or code to cleanup the old records. The good thing is that the solution can be done very specific to the needs of the people, but the bad is that in general it's toil for the community.
2. We could provide a cleanup function (--cleanup-txt-record-regexp="..") that matches only the value of "ownership" by regexp and users can use it to cleanup TXT ownership records.

The problem with migration and automatic deletion is that you do not want to "import" ownership of records you did not own before, nor you want to loose ownership of old records. The task is not easy and I tend to the suggested decision that we merge the PR and let the community cleanup their records, because `--cleanup-txt-record-regexp=".."` is likely done in a couple of lines in shell script or whatever you prefer.

In the end the txt-registry helped a lot of organizations and it's the best we have. I know people say that it's not good for diverse set of reasons, but at least it works for all providers and it also works at scale for huge orgs like my employer pretty well. I think we had close to 0 incidents caused by external-dns.

I will start reading the PR soon.

szuecs commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+1. TXT record with prefix based on requirements. Example `--txt-prefix="%{record_type}-abc-"` or `--txt-prefix="%{record_type}.abc-"`
+2. TXT record with suffix based on requirements. Example `--txt-suffix="-abc-%{record_type}"` or `--txt-suffix="-abc.%{record_type}."`
+
+Example when configured `--txt-prefix="%{record_type}-abc-"` for apex domain `example.com` the expected result is
```
when -> if

ivankatliarchuk replied:
fixed

szuecs commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+|   `a-abc-nginx-v2.ex.com.`    |  `TXT`   |
+| `nginx-v2.ex.com.`  | `CNAME`  |
+
+And when configured `--txt-suffix="-abc.%{record_type}"` for apex domain `example.com` the expected result is
```
when -> if

ivankatliarchuk replied:
fixed

szuecs commented on pkg/apis/externaldns/types.go:
```diff
@@ -615,7 +615,7 @@ func App(cfg *Config) *kingpin.Application {
 	app.Flag("txt-encrypt-enabled", "When using the TXT registry, set if TXT records should be encrypted before stored (default: disabled)").BoolVar(&cfg.TXTEncryptEnabled)
 	app.Flag("txt-encrypt-aes-key", "When using the TXT registry, set TXT record decryption and encryption 32 byte aes key (required when --txt-encrypt=true)").Default(defaultConfig.TXTEncryptAESKey).StringVar(&cfg.TXTEncryptAESKey)
-	app.Flag("txt-new-format-only", "When using the TXT registry, only use new format records which include record type information (e.g., prefix: 'a-'). Reduces number of TXT records (default: disabled)").BoolVar(&cfg.TXTNewFormatOnly)
+	app.Flag("txt-new-format-only", "Was used during TXT format migration. Enabling or disabling it will not take any effect. (deprecated)").BoolVar(&cfg.TXTNewFormatOnly)
```
IMO should be dropped same reason as the other

szuecs commented on registry/txt_utils_test.go:
```diff
@@ -0,0 +1,51 @@
+
+/**
+
+helper methods
```
text_utils filename says this already, I think we can drop the comment

ivankatliarchuk replied:
removed

szuecs commented on scripts/aws-cleanup-legacy-txt-records.py:
```diff
@@ -0,0 +1,213 @@
+    parser.add_argument("--batch-delete-count", type=int, required=False, default=2, help="Number of items to delete in single DELETE batch. Default to 2")
+    parser.add_argument("--run", action="store_true", help="Execute the cleanup. The tool will do a dry-run if --run is not specified.")
+
+    print("Run this script at your own RISKS!!!!")
```
Should this request input from the user?

szuecs commented:
I left a couple of comments but in total I lgtm

ivankatliarchuk commented:
/unhold

ivankatliarchuk commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+
+- Creates a TXT record with record type information (e.g., 'a-' prefix for A records)
+
+The TXT registry would try to guarantee a consistency in between providers and sources, if provider supports the behaviour
```
fixed

ivankatliarchuk commented on docs/flags.md:
```diff
@@ -161,7 +161,7 @@
 | `--[no-]txt-encrypt-enabled` | When using the TXT registry, set if TXT records should be encrypted before stored (default: disabled) |
 | `--txt-encrypt-aes-key=""` | When using the TXT registry, set TXT record decryption and encryption 32 byte aes key (required when --txt-encrypt=true) |
-| `--[no-]txt-new-format-only` | When using the TXT registry, only use new format records which include record type information (e.g., prefix: 'a-'). Reduces number of TXT records (default: disabled) |
+| `--[no-]txt-new-format-only` | Was used during TXT format migration. Enabling or disabling it will not take any effect. (deprecated) |
```
flag remvoved

ivankatliarchuk commented on registry/txt.go:
```diff
@@ -74,6 +72,10 @@ func NewTXTRegistry(provider provider.Provider, txtPrefix, txtSuffix, ownerID st
 	}
 
+	if newFormatOnly {
+		log.Warn("--txt-new-format-only is left for backward compatibility, it will be removed in future releases")
```
removed

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+
+The TXT registry would try to guarantee a consistency in between providers and sources, if provider supports the behaviour.
+
+If you are dealing with APEX domains, example `example.com` and TXT records are failing to be created for managed record types specified by `--managed-record-types`, consider following options:
```
```suggestion
If you are dealing with APEX domains, like `example.com` and TXT records are failing to be created for managed record types specified by `--managed-record-types`, consider following options:
```

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+1. TXT record with prefix based on requirements. Example `--txt-prefix="%{record_type}-abc-"` or `--txt-prefix="%{record_type}.abc-"`
+2. TXT record with suffix based on requirements. Example `--txt-suffix="-abc-%{record_type}"` or `--txt-suffix="-abc.%{record_type}."`
+
+Example if configured `--txt-prefix="%{record_type}-abc-"` for apex domain `ex.com` the expected result is
```
```suggestion
If configured `--txt-prefix="%{record_type}-abc-"` for apex domain `ex.com` the expected result is
```

mloiseleur commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+| `cname-a-abc-nginx-v2.ex.com.` |  `TXT`  |
+|       `nginx-v2.ex.com.`       | `CNAME` |
+
+And if configured `--txt-suffix="-abc.%{record_type}"` for apex domain `ex.com` the expected result is
```
```suggestion
If configured `--txt-suffix="-abc.%{record_type}"` for apex domain `ex.com` the expected result is
```

mloiseleur commented:
/lgtm
I left some minor suggestion.
I let @szuecs proceed with the final review.

szuecs commented on docs/registry/txt.md:
```diff
@@ -3,8 +3,51 @@
+|              Name              |  TYPE   |
+|:------------------------------:|:-------:|
+| `cname-nginx-v2-abc.a.ex.com.` |  `TXT`  |
+|      `nginx-v3.ex.com..`       | `CNAME` |
```
There are two dots. Mistake or on purpose?

ivankatliarchuk replied:
Mistake, fixed

ivankatliarchuk commented on scripts/aws-cleanup-legacy-txt-records.py:
```diff
@@ -0,0 +1,215 @@
+    parser = argparse.ArgumentParser(description="Cleanup legacy TXT records")
+    parser.add_argument("--zone-id", type=str, required=True, help="Hosted Zone ID for which to run a cleanup.")
+    parser.add_argument("--record-match", type=str, required=True, help="Record to match specific value. Example 'external-dns/owner=default'")
+    parser.add_argument("--total-items", type=int, required=False, default=100, help="Number of items to delete. Default to 10")
```
Nice one. Reduced value to 10

mloiseleur commented:
/lgtm

ivankatliarchuk commented:
![Screenshot 2025-06-25 at 10 39 05](https://github.com/user-attachments/assets/22ef7723-450a-4bd5-a994-9b92420f30a8)

frittentheke commented:
> This is a breaking change as its causing the error
> 
> ```
> flag parsing error: unknown long flag '--txt-new-format-only'
> ```
> 
> Just mentioning it as its not listed as a breaking change on the release notes. Only as a feature.

Would also be nice to point people (again) at the required steps to clean up their records.

Phil1602 commented:
> > This is a breaking change as its causing the error
> > ```
> > flag parsing error: unknown long flag '--txt-new-format-only'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Just mentioning it as its not listed as a breaking change on the release notes. Only as a feature.
> 
> Would also be nice to point people (again) at the required steps to clean up their otherwise now orphaned records in the old format.

Yep. Just opened https://github.com/kubernetes-sigs/external-dns/issues/5868 therefore.
